packages/panzer/disc-fe/src/Panzer_IntegrationValues2.cpp

```diff
@@ -746,56 +621,228 @@ generateSurfaceCubatureValues(const PHX::MDField<Scalar,Cell,NODE,Dim>& in_node_
+
+          // Normalize the binormal vector
+          const Scalar mag_b = std::sqrt(PANZER_DOT(b,b));
+          b[0] /= mag_b;
+          b[1] /= mag_b;
+          b[2] /= mag_b;
+          
         }
+        
+      }
 
-        convertNormalToRotationMatrix(normal,transverse,binormal);
+      // Now that we have a reference coordinate plane in which to align our points
+      // Point on the transverse/binormal plane
+      Scalar p0[2] = {0};
+      Scalar p1[2] = {0};
+
+      // Differential position in mesh
+      Scalar x0[3] = {0};
+      Scalar x1[3] = {0};
+      
+      // Iterate through points in cell 1 and find which point they align with in cell 0
+      for(int face_point_1=0; face_point_1<num_points_per_face; ++face_point_1){
+
+        // Get the point index in the 1 cell
+        const int point_1 = lidx_1*num_points_per_face + face_point_1;
+
+        // Load position shifted by face center
+        for(int dim=0; dim<cell_dim; ++dim)
+          x1[dim] = ip_coordinates(cell_1,point_1,dim) - xc1[dim];
+
+        // Convert position to transverse/binormal plane
+        p1[0] = PANZER_DOT(x1,t);
+        p1[1] = PANZER_DOT(x1,b);
+
+        // Set the default for the point order
+        point_order[face_point_1] = face_point_1;
+        
+        // Compare to points on the other surface
+        for(int face_point_0=0; face_point_0<num_points_per_face; ++face_point_0){
+
+          // Get the point index in the 0 cell
+          const int point_0 = lidx_0*num_points_per_face + face_point_0;
+
+          // Load position shifted by face center
+          for(int dim=0; dim<cell_dim; ++dim)
+            x0[dim] = ip_coordinates(cell_0,point_0,dim) - xc0[dim];
+
+          // Convert position to transverse/binormal plane
+          p0[0] = PANZER_DOT(x0,t);
+          p0[1] = PANZER_DOT(x0,b);
+
+          // Find the distance squared between p0 and p1
+          const Scalar p012 = (p0[0]-p1[0])*(p0[0]-p1[0]) + (p0[1]-p1[1])*(p0[1]-p1[1]);
+
+          // If the distance, compared to the size of the cell, is small, we assume these are the same points
+          if(p012 / r2 < 1.e-12){
+            point_order[face_point_1] = face_point_0;
+            break;
+          }
+
+          // Big problem - there wan't a point that aligned properly
+          TEUCHOS_ASSERT(face_point_0 != num_points_per_face-1);
 
-        for(int dim=0; dim<3; ++dim){
-          surface_rotation_matrices(cell,point,0,dim) = normal[dim];
-          surface_rotation_matrices(cell,point,1,dim) = transverse[dim];
-          surface_rotation_matrices(cell,point,2,dim) = binormal[dim];
+        }
+      }
+      
+      // Now re-order the points on face 1 to correct the alignment
+      const int point_offset = lidx_1*num_points_per_face;
+      for( int face_point_1 = 0; face_point_1 < num_points_per_face - 1; ++face_point_1 ){
+        // While the point is not yet in place - keep swapping until it is in place (N^2 operations - not great)
```

Comment:
I don't think the situation is as bad as the comment here suggests.  The point permutations are not arbitrary; in 2D I believe one is just the reverse of the other (or they are already in the sameorder).  If the points are just in reverse order, then we only need one iteration of the while loop for every two points; once we get halfway through the points, they'll be in order.  So for that case it's N/2 passes through the while loop.

In 3D, I'm not sure what all the possibilities are, but I suspect something similar holds so that the practical worst case is still O(N).

Reply:
Other than that one correction, looks good to me!